решения 5го урока

KuzinaMari opened the pull request.

KuzinaMari committed:
Решен урок 1

KuzinaMari committed:
Решен урок 1

KuzinaMari committed:
Решен урок 2 (часть 1)

KuzinaMari committed:
Решен урок 2 (часть 1)

KuzinaMari committed:
Урок 2, часть 1, исправлены ошибки

KuzinaMari committed:
Урок 2, часть 1, исправлена ошибка в задаче "normal" и решен Урок 2, часть 2

KuzinaMari committed:
Исправлены ошибки
Решен Урок 3

KuzinaMari committed:
Исправлены отступы в Уроке 3

KuzinaMari committed:
Исправлены ошибки в Уроке 3

KuzinaMari committed:
Исправлены ошибки в Уроке 3 (тривиальная, синус, косинус)

KuzinaMari committed:
Сделан Урок№4

KuzinaMari committed:
Исправлены две последние задачи в Уроке№4

KuzinaMari committed:
Исправлена russian в Уроке№4

KuzinaMari committed:
Сделан Урок№5

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,37 @@ fun main(args: Array<String>) {
+                "сентября" -> "09"
+                "октября" -> "10"
+                "ноября" -> "11"
+                "декабря" -> "12"
```
В таких случаях обычно ПЕРЕД функцией заводится переменная, содержащая список названий месяцев:

```
val MONTH_NAMES = listOf("января", "февраля", "марта", ...)
```

после чего вы можете использовать `MONTH_NAMES.indexOf(parts[1]) + 1` для определения нужного вам номера месяца. Если результат `indexOf` равен -1 -- значит, название не соответствует ни одному из месяцев.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -69,7 +98,35 @@ fun dateStrToDigit(str: String): String = TODO()
+                "09" -> "сентября "
+                "10" -> "октября "
+                "11" -> "ноября "
+                "12" -> "декабря "
```
А здесь вы можете использовать обращение по индексу к тому же массиву `MONTH_NAMES`.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -69,7 +98,35 @@ fun dateStrToDigit(str: String): String = TODO()
+                else -> return ""
+            }
+
+            result = result + parts[2].toInt().toString()
```
Почему не просто `result += parts[2]` ?

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,37 @@ fun main(args: Array<String>) {
+    if (parts.size == 3) {
+        try {
+            if (parts[0].toInt() in 0..9) result = '0' + parts[0].toInt().toString() + "."
+            else result = parts[0].toInt().toString() + "."
```
Здесь и далее: `string.toInt().toString()` даёт тот же самый `string`, который был на входе

mglukhikh replied:
Это замечание остаётся в силе

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +140,37 @@ fun dateDigitToStr(digital: String): String = TODO()
+    var result: String = ""
+    if (phone[0].toString() == "+") {
+        result += "+"
+        for (i in 1..phone.length - 1) {
```
Код этого цикла `for` здесь и ниже повторяется. Подумайте, как избежать повторения (можно его вынести в отдельную функцию или переписать код в этой так, чтобы дублирования не было).

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +209,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+        string = jumps.split(" ")
+        if (string.size % 2 == 1) return -1
+        for (i in 0..string.size - 1 step 2) {
+            if ((string[i].toInt() > maxMax) && (checkPlus(string[i + 1]) == true)) {
```
Запись `someBoolean == true` в условии избыточна, можно (и нужно!) писать просто `someBoolean`.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +245,28 @@ fun bestHighJump(jumps: String): Int = TODO()
+    var result = 0
+    try {
+        if ((parts[0] == "-") || (parts[0] == "+"))
+            for (i in 1..parts.size - 1 step 2) if (parts[i - 1] == "-") result -= parts[i].toInt()
```
Здесь лучше поставить фигурные скобки хотя бы для `for`, иначе непонятно, что к чему относится:

```
for (...) {
    if (...) result -= ...
    else result += ...
}
```

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +245,28 @@ fun bestHighJump(jumps: String): Int = TODO()
+            else result += parts[i].toInt()
+        else {
+            result += parts[0].toInt()
+            for (i in 2..parts.size - 1 step 2) if (parts[i - 1] == "-") result -= parts[i].toInt()
```
Код повторяется, подумайте, можно ли избежать повтора

mglukhikh commented:
recheck all

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +403,147 @@ fun roman(n: Int): String = TODO()
+
+    }
+    when {
+        (thousands100 == 1) -> result += "сто"
```
Здесь и далее, следовало записать константы `"сто"`, `"двести"` и последующие в списки и затем использовать обращение по индексу к этим спискам. Лучше определить их на верхнем уровне (перед функцией `russian`), например:

```
val HUNDREDS_LIST = listOf("сто", "двести", ...)
```

mglukhikh replied:
Это замечание остаётся в силе

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -214,7 +320,81 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
+        thousands --
+    }
+    when {
+        (hundreds == 1) -> result += ("C")
```
Логика этих трёх `when` одинакова. Подумайте, как можно избежать дублирования кода

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +140,37 @@ fun dateDigitToStr(digital: String): String = TODO()
-fun flattenPhoneNumber(phone: String): String = TODO()
+fun flattenPhoneNumber(phone: String): String {
+    var result: String = ""
+    if (phone[0].toString() == "+") {
```
Исходный номер может оказаться пустой строкой (см. сообщение бота)

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +166,19 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+    var i = 0
+    var sinus: Double = x%(Math.PI*2)
+    var number: Double = x%(Math.PI*2)
+    while (Math.abs(number) * 1000 > eps) {
```
У вас не проходят тесты бота для задач `sin` и `cos`. Проблема связана с понижением точности расчёта при увеличении `x` из-за увеличения необходимого числа членов ряда. Вы использовали в целом правильную идею (следует с самого начала уменьшить `x`, используя периодичность синуса), но неверно её реализовали. Прочитайте код внимательно и исправьте ошибки. NB: умножение на 1000 в цикле while ничего хорошего вам не даст.

mglukhikh reviewed (changes requested):
В целом очень хорошо. Однако, три задачи решены неверно (sin и cos из 3-го урока и flattenPhoneNumber из 4-го). Кроме этого, в задачах 4-го и 5-го урока вы часто дублируете код в ситуации, где можно было использовать списки. Прочитайте, пожалуйста, мои замечания и исправьте ошибки.

KuzinaMari committed:
Исправлены ошибки в Уроке№5, переписаны задачи Sin и Cos (Урок№3)
/Использую String/toInt().toString() для того, чтобы избавиться от впереди стоящего нуля, например, когда в исходной строке 03.10.2016, а требуется вывести 3 ноября 2016/

KuzinaMari committed:
Урок№5, задача dateStrToDigit

KuzinaMari committed:
Урок№5, задача dateStrToDigit

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -129,7 +256,16 @@ fun plusMinus(expression: String): Int = TODO()
  */
-fun firstDuplicateIndex(str: String): Int = TODO()
+fun firstDuplicateIndex(str: String): Int {
+    var string = str.toLowerCase().split(" ")
```
Очень неудачное название переменной. Любому читателю будет очевидно, что переменная с именем `string` имеет тип `String`, хотя в вашем случае это совершенно неверно -- её тип `List<String>`. Можно назвать её `parts` или `stringParts`, или ещё как-нибудь в таком духе.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +219,33 @@ fun bestHighJump(jumps: String): Int = TODO()
+    var j = 0
+    for (i in 0..expression.length - 1)
+        if ((expression == "") || ((expression[i] != '+') && (expression[i] != '-') && (expression[i] != ' ') && (expression[i] !in '0'..'9')))
+            return throw IllegalArgumentException("Description")
```
`return` здесь не нужен, IDEA вам его подсветит как недостижимый. Просто: `throw ...`.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +219,33 @@ fun bestHighJump(jumps: String): Int = TODO()
+fun plusMinus(expression: String): Int {
+    var j = 0
+    for (i in 0..expression.length - 1)
+        if ((expression == "") || ((expression[i] != '+') && (expression[i] != '-') && (expression[i] != ' ') && (expression[i] !in '0'..'9')))
```
Сравнение выражения с пустой строкой лучше вынести из цикла и записать как `expression.isEmpty()`, это быстрее, чем сравнение строк. Что касается цикла, то лучше записать его через перебор символов `for (char in expression)` и поставить фигурные скобки -- это улучшит читаемость

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +219,33 @@ fun bestHighJump(jumps: String): Int = TODO()
+    for (i in 0..expression.length - 1)
+        if ((expression == "") || ((expression[i] != '+') && (expression[i] != '-') && (expression[i] != ' ') && (expression[i] !in '0'..'9')))
+            return throw IllegalArgumentException("Description")
+    var resultExpression = expression
```
Не понял назначение этой переменной. Чем плох просто `expression`?

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +219,33 @@ fun bestHighJump(jumps: String): Int = TODO()
+        }
+
+    } catch (e: IllegalArgumentException) {
+        throw IllegalArgumentException()
```
В таком виде блок `catch` не имеет особого смысла, так как бросает дальше то же самое исключение, которое ловит. Возможно, поймать вы хотели `NumberFormatException` (и это как раз было бы правильно, так как контролировало бы нарушение формата строки)

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +183,32 @@ fun bestLongJump(jumps: String): Int = TODO()
  * При нарушении формата входной строки вернуть -1.
  */
-fun bestHighJump(jumps: String): Int = TODO()
+fun checkPlus(str: String): Boolean {
```
В качестве упражнения: попробуйте написать эту функцию с помощью функции высшего порядка `any` (это её сократит примерно в 5 раз)

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +183,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+    if (jumps == "") return -1
+    var maxMax = -1
+    var maxI = -1
+    var string = listOf<String>()
```
`stringList` или `strings`, но никак не `string`, поскольку переменная с названием `string` с очевидностью должна иметь тип `String`.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +183,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+    var maxI = -1
+    var string = listOf<String>()
+    try {
+        string = jumps.split(" ")
```
Лучше было бы определить переменную прямо здесь: `val stringList = ...`.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +156,22 @@ fun flattenPhoneNumber(phone: String): String = TODO()
-fun bestLongJump(jumps: String): Int = TODO()
+fun bestLongJump(jumps: String): Int {
+    for (i in 0..jumps.length - 1)
+        if ((jumps == "")
```
Эту проверку лучше вынести из цикла

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +156,22 @@ fun flattenPhoneNumber(phone: String): String = TODO()
  */
-fun bestLongJump(jumps: String): Int = TODO()
+fun bestLongJump(jumps: String): Int {
+    for (i in 0..jumps.length - 1)
```
Лучше так: `for (jump in jumps) { ... }`

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +156,22 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    var parts0 = listOf<Int>()
+    for (i in 0..parts.size - 1) if (parts[i] != "") parts0 = parts0 + parts[i].toInt()
+    var result = -1
+    for (i in 0..parts0.size - 1) result = Math.max(result, parts0[i])
```
Здесь всё получилось как-то сложно. Вначале вы выкинули из строки чёрточки и проценты, а потом получили проблему с возможными пустыми строками. Кроме этого, вполне можно обойтись без списка `parts0` и без двух отдельных циклов.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +123,28 @@ fun dateDigitToStr(digital: String): String = TODO()
+fun flattenPhoneNumber(phone: String): String {
+    val result = phone.filter { (it in '0'..'9') || (it == '+') }
+
+    for (i in 0..phone.length - 1)
```
`for (char in phone) { ... }`, и никаких индексов не нужно

mglukhikh replied:
`for (char in phone) { ... }`, и никаких индексов не нужно

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +123,28 @@ fun dateDigitToStr(digital: String): String = TODO()
+        if (result[0] == '+') {
+            for (i in 1..result.length - 1) if (result[i] !in '0'..'9') return ""
+        } else {
+            for (i in 0..result.length - 1) if (result[i] !in '0'..'9') return ""
```
Эти две ветки очень похожи, необязательно их повторять полностью

mglukhikh replied:
Эти две ветки очень похожи, необязательно их повторять полностью

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,31 @@ fun main(args: Array<String>) {
+            if (parts[0].toInt() in 0..9) result = '0' + parts[0].toInt().toString() + "."
+            else result = parts[0].toInt().toString() + "."
+            val month1 = MONTH_NAME.indexOf(parts[1]) + 1
+            if ((month1 != 0) && (month1 < 10)) {
```
Здесь можно использовать `String.format("%02d", month1)` для получения номера месяца из двух цифр

mglukhikh reviewed (changes requested):
Ага, спасибо, тесты вы поправили, остались только замечания по стилю.

KuzinaMari committed:
Исправлены ошибки в Уроке№5

KuzinaMari committed:
Исправлены ошибки в Уроке№4

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -2,6 +2,8 @@
 
 package lesson5.task1
 
+import kotlin.system.exitProcess
```
Это лишнее

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -157,19 +159,18 @@ fun flattenPhoneNumber(phone: String): String {
-    for (i in 0..parts.size - 1) if (parts[i] != "") parts0 = parts0 + parts[i].toInt()
     var result = -1
-    for (i in 0..parts0.size - 1) result = Math.max(result, parts0[i])
+    for (i in 0..parts.size - 1)
```
Лучше `for (part in parts) { ... }`

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -319,81 +319,24 @@ fun decimalFromString(str: String, base: Int): Int {
  */
+val UNITS1_LIST = listOf<String>("", "I", "II", "III", "VI", "V", "IV", "IIV", "IIIV", "XI")
+val DOZENS1_LIST = listOf<String>("", "X", "XX", "XXX", "LX", "L", "XL", "XXL", "XXXL", "CX")
+val HUNDREDS1_LIST = listOf<String>("", "C", "CC", "CCC", "DC", "D", "CD", "CCD", "CCCD", "MC")
```
Лучше убрать 1 из всех этих имён. Кроме этого, не DOZENS (дюжины), а TENS (десятки)

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -403,6 +346,11 @@ fun roman(n: Int): String {
+val UNITS_LIST = listOf("один", "два", "три", "четыре", "пять", "шесть", "семь", "восемь", "девять")
+val THOUSANDS_LIST = listOf("одна", "две")
+val DOSENS1_LIST = listOf("десять", "одиннадцать", "двенадцать", "тринадцать", "четырнадцать", "пятнадцать", "шестнадцать", "семнадцать", "восемнадцать", "девятнадцать")
+val DOSENS_LIST = listOf("двадцать", "тридцать", "сорок", "пятьдесят", "шестьдесят", "семьдесят", "восемьдесят", "девяносто")
```
Здесь я бы использовал имена `TEENS_LIST` и `TENS_LIST`.

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -319,81 +319,24 @@ fun decimalFromString(str: String, base: Int): Int {
+    val n3 = n2.toString().reversed() //в n3 лежит изначально введенное число без тысяч, переведенное в строку и перевернутое//
+    if (n3.length >= 1) result2 += UNITS1_LIST[n3[0].toString().toInt()] //если число без тысяч != 0, то берем последнюю цифру, записываем римскими/
+    if (n3.length >= 2) result2 += DOZENS1_LIST[n3[1].toString().toInt()]//если в числе без тысяч есть десятки, то зайдет сюда и запишет римскими//
+    if (n3.length >= 3) result2 += HUNDREDS1_LIST[n3[2].toString().toInt()]//если есть и сотни, то тоже запишет римскими//
```
`n3[i] - '0'` было бы понятнее, чем `n3[i].toString().toInt()`.

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -436,53 +384,53 @@ fun russian(n: Int): String {
+        (thousands100 == 6) -> result += HUNDREDS_LIST[5]
+        (thousands100 == 7) -> result += HUNDREDS_LIST[6]
+        (thousands100 == 8) -> result += HUNDREDS_LIST[7]
+        (thousands100 == 9) -> result += HUNDREDS_LIST[8]
```
Мне кажется, эти 9 строчек можно превратить в одну

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -436,53 +384,53 @@ fun russian(n: Int): String {
+        (thousands10 == 6) -> result += DOSENS_LIST[4]
+        (thousands10 == 7) -> result += DOSENS_LIST[5]
+        (thousands10 == 8) -> result += DOSENS_LIST[6]
+        (thousands10 == 9) -> result += DOSENS_LIST[7]
```
Аналогично

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -436,53 +384,53 @@ fun russian(n: Int): String {
+        ((thousands10 != 1) && (thousands == 6)) -> result += UNITS_LIST[5]
+        ((thousands10 != 1) && (thousands == 7)) -> result += UNITS_LIST[6]
+        ((thousands10 != 1) && (thousands == 8)) -> result += UNITS_LIST[7]
+        ((thousands10 != 1) && (thousands == 9)) -> result += UNITS_LIST[8]
```
Аналогично (здесь, правда, скорее в две)

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -495,48 +443,48 @@ fun russian(n: Int): String {
+        (hundreds == 6) -> result += HUNDREDS_LIST[5]
+        (hundreds == 7) -> result += HUNDREDS_LIST[6]
+        (hundreds == 8) -> result += HUNDREDS_LIST[7]
+        (hundreds == 9) -> result += HUNDREDS_LIST[8]
```
Аналогично

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -495,48 +443,48 @@ fun russian(n: Int): String {
+        (dozens == 6) -> result += DOSENS_LIST[4]
+        (dozens == 7) -> result += DOSENS_LIST[5]
+        (dozens == 8) -> result += DOSENS_LIST[6]
+        (dozens == 9) -> result += DOSENS_LIST[7]
```
Аналогично.

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -495,48 +443,48 @@ fun russian(n: Int): String {
+        ((dozens != 1) && (units == 6)) -> result += UNITS_LIST[5]
+        ((dozens != 1) && (units == 7)) -> result += UNITS_LIST[6]
+        ((dozens != 1) && (units == 8)) -> result += UNITS_LIST[7]
+        ((dozens != 1) && (units == 9)) -> result += UNITS_LIST[8]
```
Аналогично

mglukhikh reviewed (changes requested):
Стало лучше, но ещё, как видите, есть над чем работать. Особенно это касается функции `russian`.

KuzinaMari committed:
Сделаны исправления (russian и Урок№5)

KuzinaMari committed:
Исправлена ошибка в russian

KuzinaMari committed:
Исправлена ошибка в russian

mglukhikh commented:
recheck all

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -1,7 +1,9 @@
 package lesson4.task1
 
 import lesson1.task1.discriminant
+import java.util.function.IntToDoubleFunction
```
Используете ли вы эту функцию? Если нет, то импорт нужно удалить

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh replied:
Используете ли вы эту функцию? Если нет, то импорт нужно удалить

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,43 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+    for (i in 0..v.size - 1) {
+        val element = v[i] * v[i]
+        result.add(element)
+    }
```
For-each: `for (element in v) { result.add(element * element) }`. Ещё короче это можно записать с помощью функции высшего порядка `map`.

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +235,24 @@ fun factorizeToString(n: Int): String = TODO()
+            result.add(i)
+        }
+        result.add(nn)
+    } else
```
Действительно ли нужна эта ветка `else`? Впечатление, что она полезна только для случая с `n = 0`. А его можно обработать проще в начале функции

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +262,19 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    result = convert(n, base)
+    for (i in 0..result.size - 1) {
+        if (result[i] > 9) {
+            result1 += (result[i] + 87).toChar()
```
Что такое 87? Замените это на соответствующий символ

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +283,15 @@ fun convertToString(n: Int, base: Int): String = TODO()
+    var result = 0
+    var numbers = digits.reversed()
+    for (i in 0..digits.size - 1) {
+        val ss = (numbers[i] * Math.pow(base.toDouble(), i.toDouble())).toInt()
```
Здесь лучше написать свою функцию `pow` для целых чисел, а не заниматься конвертацией вначале в `Double`, а потом в `Int`. Помимо того, что некрасиво, вы можете здесь нарваться на ошибку округления

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh replied:
Здесь лучше написать свою функцию `pow` для целых чисел, а не заниматься конвертацией вначале в `Double`, а потом в `Int`. Помимо того, что некрасиво, вы можете здесь нарваться на ошибку округления

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +302,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+    var string1 = str
+    for (i in 0..string1.length - 1)
+        if (string1[i] in '0'..'9') list += ((string1[i]).toInt() - 48)
+        else list += ((string1[i]).toInt() - 87)
```
48, 87 -- что они значат? Подставьте вместо них соответствующие символы

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +302,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+    var list: List<Int>
+    list = listOf()
+    var string1 = str
+    for (i in 0..string1.length - 1)
```
Здесь лучше использовать for-each: `for (char in srting1) { ... }`.

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh replied:
Здесь лучше использовать for-each: `for (char in srting1) { ... }`.

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +302,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    var list: List<Int>
+    list = listOf()
+    var string1 = str
```
Зачем `string1`, чем вас `str` не устраивает?

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh replied:
Зачем `string1`, чем вас `str` не устраивает?

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -1,6 +1,9 @@
+
 package lesson5.task1
 
+import kotlin.system.exitProcess
```
Мне кажется, этот импорт не нужен

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +185,28 @@ fun bestLongJump(jumps: String): Int = TODO()
+                maxI = i
+            }
+        }
+        if (maxI == -1) return -1 else return stringList[maxI].toInt()
```
Здесь стоило использовать `maxMax`, раз уж вы её создали. И можно обойтись совсем без переменной `maxI`

mglukhikh replied:
Это замечание всё ещё в силе

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +185,28 @@ fun bestLongJump(jumps: String): Int = TODO()
  * При нарушении формата входной строки вернуть -1.
  */
-fun bestHighJump(jumps: String): Int = TODO()
+fun checkPlus(str: String): Boolean = str.any { str == "+" }
```
`str.any { it == "+" }`. `str` -- это название параметра `checkPlus`. А `it` -- это название единственного параметра лямбды (в том случае, если оно не указано явно, как например здесь: `str.any { char -> char == "+" }`). По-видимому, из-за этой ошибки у вас сейчас не проходят тесты `bestHighJump`

mglukhikh replied:
Замечание более не актуально

mglukhikh commented on test/lesson4/task1/Tests.kt:
```diff
@@ -180,17 +180,22 @@ class Tests {
+        assertEquals("триста десять", russian(310))
         assertEquals("двадцать две тысячи девятьсот шестьдесят четыре", russian(22964))
-        assertEquals("сто девятнадцать тысяч пятьсот восемь", russian(119508))
+        assertEquals("сто одиннадцать тысяч пятьсот восемь", russian(111508))
```
На будущее: никогда не исправляйте имеющиеся тесты на новые. Лучше оставить старый тест, и добавить новый. Тогда вы проверите оба случая

mglukhikh reviewed (changes requested):
Замечания, к моему сожалению, по-прежнему остаются

KuzinaMari committed:
Выполнен Урок№ 6 (Task 1, Task 2)

KuzinaMari committed:
Выполнен Урок№ 6 (Task 1, Task 2)

KuzinaMari committed:
Выполнен Урок№ 6 (Task 1, Task 2)

KuzinaMari committed:
исправлены ошибки в уроке 6, но в двух задачах не совпадают углы(после 11го знака после запятой), не понимаю как это исправить
выполнен урок 7.1

KuzinaMari commented:
Извините пожалуйста, совсем не понимаю, почему в Уроке 6.2 не проходят проверку задачи. Исключение не ловится? Получается нужен catch..?

mglukhikh commented:
Вы кидаете не то исключение. Нужно IllegalArgument..., а вы кидаете IllegalAccess...

KuzinaMari committed:
исправлены исключения

mglukhikh commented:
recheck all

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,75 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun digitNumber(n: Int): Int = TODO()
+fun digitNumber(n: Int): Int {
+    var number: Int = 0
+    var n2: Int = n
```
Тип локальных переменных указывать не требуется

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,75 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun minDivisor(n: Int): Int = TODO()
+fun minDivisor(n: Int): Int {
+    var i: Int = 2
```
Тип локальных переменных указывать не требуется

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,75 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun maxDivisor(n: Int): Int = TODO()
+fun maxDivisor(n: Int): Int {
+    var count: Int = 1
```
Тип локальных переменных указывать не требуется

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +135,15 @@ fun maxDivisor(n: Int): Int = TODO()
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
+    val nn = Math.max(m, n)
+    var failed: Boolean = false
```
Тип локальных переменных указывать не требуется

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +135,15 @@ fun maxDivisor(n: Int): Int = TODO()
+    for (i in 2..nn) {
+        if ((n % i == 0) && (m % i == 0)) failed = true
+    }
+    if (failed == true) return false else return true
```
Эту чудо-конструкцию можно записать гораздо короче

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -105,7 +152,12 @@ fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+
+fun squareBetweenExists(m: Int, n: Int): Boolean =
+        if (m <= sqr(Math.sqrt(n.toDouble()).toInt().toDouble()) && sqr(Math.sqrt(n.toDouble()).toInt().toDouble()) <= n) true
+        else false
```
Вспомните то, что рассказывал на последней лекции Марат Халимович, и напишите эту функцию нормально. Выглядит это абсолютно страшно

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +212,33 @@ fun revert(n: Int): Int = TODO()
+        result = result * 10 + nn % 10
+        nn /= 10
+    }
+    if (result == n) return true else return false
```
Эта функция очень похожа на `revert`. Попробуйте написать одно через другое

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +212,33 @@ fun revert(n: Int): Int = TODO()
+        if (nn % 10 != nn1) indicator = false
+        nn /= 10
+    }
+    if (indicator == false) return true else return false
```
Эту конструкцию можно записать гораздо короче

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +212,33 @@ fun revert(n: Int): Int = TODO()
+fun hasDifferentDigits(n: Int): Boolean {
+    var nn1 = n % 10
+    var nn = n
+    var indicator = true
```
Зачем нужна эта переменная?

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,43 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+fun center(list: MutableList<Double>): MutableList<Double> {
+    val count = list.size
+    val sum = list.sum()
+    val arithmeticaverage = sum / count
```
`camelCase`: `arithmeticAverage`. А ещё лучше просто `mean`. И не забывайте, что выше вы уже написали эту функцию

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -148,7 +191,15 @@ fun polynom(p: List<Double>, x: Double): Double = TODO()
+fun accumulate(list: MutableList<Double>): MutableList<Double> {
+
+    var length = 0.0
+    for (i: Int in 0..list.size - 1) {
```
Указание индекса здесь не нужно

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -213,8 +319,25 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
+val HUNDREDS_LIST = listOf<String>("", "C", "CC", "CCC", "DC", "D", "CD", "CCD", "CCCD", "MC")
+fun roman(n: Int): String {
+    var result: String = ""
+    var result2: String = ""
```
Типы указывать не требуется. Неудачные названия переменных

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -213,8 +319,25 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
-fun roman(n: Int): String = TODO()
+val UNITS_LIST = listOf<String>("", "I", "II", "III", "VI", "V", "IV", "IIV", "IIIV", "XI")
+val TENS_LIST = listOf<String>("", "X", "XX", "XXX", "LX", "L", "XL", "XXL", "XXXL", "CX")
+val HUNDREDS_LIST = listOf<String>("", "C", "CC", "CCC", "DC", "D", "CD", "CCD", "CCCD", "MC")
```
Типовой параметр `String` указывать не нужно, и так ясно, что это список строк

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +346,91 @@ fun roman(n: Int): String = TODO()
+    if ((units != 0) && (tens != 1) && ((tens != 0) || (hundreds != 0) || (thousands != 0) || (thousands100 != 0) || (thousands10 != 0)))
+        result += " "
+    if ((tens != 1) && (units >= 1))
+        result += UNITS_LIST[units]
```
Данная функция состоит из двух практически повторяющихся фрагментов кода. Предлагаю вам вынести в отдельную функцию расчёт строкового представления для чисел от 1 до 999 и использовать её.

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -76,80 +84,135 @@ data class Segment(val begin: Point, val end: Point)
  */
-fun circleByDiameter(diameter: Segment): Circle = TODO()
+fun circleByDiameter(diameter: Segment): Circle
+        = Circle(Point(((diameter.begin.x + diameter.end.x) / 2), ((diameter.begin.y + diameter.end.y) / 2)), (diameter.begin.distance(diameter.end) / 2))
```
Очень длинная строка

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -76,80 +84,135 @@ data class Segment(val begin: Point, val end: Point)
+fun corner(a: Point, b: Point): Double {
+    val x = b.x - a.x
+    val y = b.y - a.y
+        return Math.atan(y / x)
```
Форматирование

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -76,80 +84,135 @@ data class Segment(val begin: Point, val end: Point)
  * Построить прямую по двум точкам
  */
-fun lineByPoints(a: Point, b: Point): Line = TODO()
+fun corner(a: Point, b: Point): Double {
```
Тот угол, который вы имели здесь в виду, по английски называется `angle`. А `corner` -- это, например, угол футбольного поля или комнаты.

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -76,80 +84,135 @@ data class Segment(val begin: Point, val end: Point)
+fun bisectorByPoints(a: Point, b: Point): Line{
+    var angle = corner(a, b) + Math.PI / 2
+    while (angle >= Math.PI){
+    angle = angle - Math.PI  }
```
Форматирование

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -76,80 +84,135 @@ data class Segment(val begin: Point, val end: Point)
+        var answer = Pair(0, 0)
+        var min = -1.0
+        var indicator = 1
+        if (circles.size >= 2) {
```
См. условие: если в списке менее двух окружностей...

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -76,80 +84,135 @@ data class Segment(val begin: Point, val end: Point)
+    fun findNearestCirclePair(vararg circles: Circle): Pair<Circle, Circle> {
+        var answer = Pair(0, 0)
+        var min = -1.0
+        var indicator = 1
```
Нельзя ли обойтись без этой переменной?

mglukhikh commented on src/lesson6/task2/Chess.kt:
```diff
@@ -23,7 +24,15 @@ data class Square(val column: Int, val row: Int) {
+    fun notation(): String {
+        if ((row in 1..8) && (column in 1..8)) {
+            var result = ""
+            val list_lat = listOf("a", "b", "c", "d", "e", "f", "g", "h")
```
`"abcdefgh"`. Не нужно создавать списки символов там, где можно обойтись строкой

mglukhikh commented on src/lesson6/task2/Chess.kt:
```diff
@@ -33,7 +42,12 @@ data class Square(val column: Int, val row: Int) {
+    if ((notation.length != 2) || (notation[0] !in 'a'..'h') || (notation[1] !in '1'..'8'))
+        throw IllegalArgumentException()
+    val list_lat = mapOf('a' to 1, 'b' to 2, 'c' to 3, 'd' to 4, 'e' to 5, 'f' to 6, 'g' to 7, 'h' to 8)
+    return Square(list_lat[notation[0]]!!, notation[1].toString().toInt())
```
Здесь всё можно сделать гораздо проще и изящнее, используя вычитание символов

mglukhikh commented on src/lesson6/task2/Chess.kt:
```diff
@@ -58,7 +72,17 @@ fun square(notation: String): Square = TODO()
+            ((start.column == end.column) && (start.row != end.row)) -> index = 1
+            ((start.column != end.column) && (start.row == end.row)) -> index = 1
+            ((start.column != end.column) && (start.row != end.row)) -> index = 2
+        }
```
`when` можно упростить. Кроме этого, поехало форматирование

mglukhikh commented on src/lesson6/task2/Chess.kt:
```diff
@@ -74,7 +98,18 @@ fun rookMoveNumber(start: Square, end: Square): Int = TODO()
-fun rookTrajectory(start: Square, end: Square): List<Square> = TODO()
+fun rookTrajectory(start: Square, end: Square): List<Square> {
+    var result = mutableListOf<Square>()
+    result.add(Square(start.column, start.row))
```
Просто `result.add(start)`

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,58 @@ interface Matrix<E> {
+        if ((other is MatrixImpl<*>) && (height == other.height) && (width == other.width))
+            for (i in 0..height - 1)
+                for (j in 0..width - 1)
+                    if (other[i, j] != this[i, j]) return false
```
Вы можете просто сравнить списки

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,58 @@ interface Matrix<E> {
+            for (i in 0..height - 1)
+                for (j in 0..width - 1)
+                    if (other[i, j] != this[i, j]) return false
+        return true
```
Этот результат неверен (вернее, верен не всегда), подумайте ещё раз, когда мы можем попасть в эту ветку

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,58 @@ interface Matrix<E> {
+        for (row in 0..height - 1) {
+            bilder.append("[")
+            for (column in 0..width - 1) {
+                bilder.append(this[row, column])
```
Вы не сможете что-либо понять в таком строковом представлении. Следует добавить какие-либо разделители между элементами строки (например, запятые или пробелы)

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,58 @@ interface Matrix<E> {
 
-    override fun toString(): String = TODO()
+    override fun toString(): String {
+        val bilder = StringBuilder()
```
`builder`

mglukhikh commented:
recheck all

mglukhikh reviewed (changes requested):
Код довольно грязный плюс неверно работает `flattenPhoneNumber`. Пожалуйста, исправьте замечания, прежде чем двигаться дальше

KuzinaMari committed:
Исправлены ошибки

mglukhikh commented:
Ваши изменения посмотрю несколько позже, но пока обратите внимание. что часть тестов не пройдена.

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +332,83 @@ fun roman(n: Int): String = TODO()
+val TENS_LIST = listOf("", "двадцать", "тридцать", "сорок", "пятьдесят", "шестьдесят", "семьдесят", "восемьдесят", "девяносто")
+val HUNDREDS_LIST = listOf("", "сто", "двести", "триста", "четыреста", "пятьсот", "шестьсот", "семьсот", "восемьсот", "девятьсот")
+
+fun less_thousand(a: Int, b: Int, c: Int, index: Int): String {
```
camelCase: `lessThousand`. Кроме этого, следует придумать более говорящие имена для параметров -- по смыслу a это сотни, b это десятки, c это единицы. Насчёт index сообразите сами.

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,52 @@ interface Matrix<E> {
+        for (row in 0..height - 1) {
+            sb.append("[")
+            for (column in 0..width - 1) {
+                sb.append(this[row, column])
```
При подобной реализации, у вас строчка матрицы из чисел 1, 23, 4, 56 будет добавлена в строку как 123456, и отдельные числа будет не различить

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,52 @@ interface Matrix<E> {
 
-    override fun equals(other: Any?) = TODO()
+    override fun equals(other: Any?) =
+            other is MatrixImpl<*> && height == other.height && width == other.width && list.equals(other.list)
```
Лучше `==` вместо `equals`

mglukhikh replied:
Также, отсутствует функция `hashCode()`

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +246,19 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    result = convert(n, base)
+    for (i in 0..result.size - 1) {
+        if (result[i] > 9) {
+            result1 += (result[i] + 'a'.toInt()).toChar()
```
Вы здесь кое-что забыли. `result[i]` равное 10 должно превратиться в `'a'`, а у вас здесь получится `'a' + 10`

mglukhikh commented on src/lesson6/task2/Chess.kt:
```diff
@@ -33,7 +43,12 @@ data class Square(val column: Int, val row: Int) {
  */
-fun square(notation: String): Square = TODO()
+fun square(notation: String): Square {
+    if ((notation.length == 2) && (notation[0] in 'a'..'h') || (notation[1] in '1'..'8' )) {
```
В этом условии напутано с логическими операциями, будьте внимательнее

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -187,4 +340,40 @@ fun fromRoman(roman: String): Int = TODO()
+fun computeDeviceCells(cells: Int, commands: String): List<Int> = TODO()
+
+
+fun myFun(coins: String, sum: Double): List<String> {
```
Пожалуйста, удалите эту функцию из проекта

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +181,27 @@ fun bestLongJump(jumps: String): Int = TODO()
  * При нарушении формата входной строки вернуть -1.
  */
-fun bestHighJump(jumps: String): Int = TODO()
+fun checkPlus(str: String): Boolean = str.any { str == "+" }
```
Эта функция возвращает `true`, если и только если `str == "+"`. Мне кажется, вы ожидали от неё другого поведения. Из-за этого, `bestHighJump` не проходит тесты

mglukhikh reviewed (changes requested):
Осталась часть старых замечаний плюс появилось несколько новых. Четыре задачи не проходят тесты

KuzinaMari committed:
Исправлены ошибки

KuzinaMari committed:
Исправленa ошибкa в bestHightJump

KuzinaMari committed:
Исправленa ошибкa в bestHightJump

mglukhikh commented:
recheck all

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +78,21 @@ fun timeForHalfWay(t1: Double, v1: Double,
+                       rookX2: Int, rookY2: Int): Int {
+    var rook1: Int = -2
+        var rook2: Int = -2
+        var indicator : Int = -2
```
Поехало форматирование

mglukhikh commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +57,14 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
+    if ((a <= r) && ((b <= s) || (c <= s))) indicator = 1
+    else if ((b <= r) && ((a <= s) || (c <= s))) indicator = 1
+    else if ((c <= r) && ((a <= s) || (b <= s))) indicator = 1
+    else indicator = 0
```
Переменная `indicator` не нужна в этой функции

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +78,21 @@ fun timeForHalfWay(t1: Double, v1: Double,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+    var rook1: Int = -2
+        var rook2: Int = -2
```
По смыслу обе переменные `rookX` -- логические (есть ли угроза от ладьи? Да/Нет), значит, их и следует сделать таковыми

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -166,4 +254,18 @@ fun squareSequenceDigit(n: Int): Int = TODO()
+
+
+fun fibSequenceDigit(n: Int): Int {
+    n.toString().length
```
Зачем нужна эта операция?

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,8 +179,26 @@ fun bestLongJump(jumps: String): Int = TODO()
+    try{
+        var jumps = jumps.split(" ")
+        if (jumps.size % 2 == 1)
+        return -1
```
Поехало форматирование

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,8 +179,26 @@ fun bestLongJump(jumps: String): Int = TODO()
+    return maxhight
+}
+catch (e: NumberFormatException) {return -1}
+}
```
И опять поехало форматирование

mglukhikh reviewed (changes requested):
Вы решили довольно много сложных задач, что хорошо, но стиль ваш не слишком хорош, что хуже. В принципе, замечаний осталось немного, вам осталось их исправить и решить что-нибудь из урока 7.2 (на четыре) и, также, из урока 8 (на пять).

KuzinaMari committed:
Исправлены ошибки
Решена задача из Урока 7(2) и Урока 8

mglukhikh commented:
recheck all

mglukhikh commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -75,7 +77,56 @@ fun generateSpiral(height: Int, width: Int): Matrix<Int> = TODO()
+    }
+}
+
+fun edge(height: Int, width: Int, streamline: Direction): Cell {
```
Здесь лучше принимать `cell: Cell` как параметр, а не `height` и `width` отдельно. Посмотрите хотя бы на места, где вы вызываете эту функцию.

mglukhikh commented on src/lesson8/task1/Files.kt:
```diff
@@ -72,9 +72,21 @@ fun countSubstrings(inputName: String, substrings: List<String>): Map<String, In
-    TODO()
+    val outputStream = File(outputName).bufferedWriter()
+    val map = mapOf<String, String>("Ы" to "И", "ы" to "и", "Я" to "А", "я" to "а", "Ю" to "У", "ю" to "у")
+    for (line in File(inputName).readLines())
```
Отсутствуют фигурные скобки

mglukhikh commented on src/lesson8/task1/Files.kt:
```diff
@@ -72,9 +72,21 @@ fun countSubstrings(inputName: String, substrings: List<String>): Map<String, In
+    val outputStream = File(outputName).bufferedWriter()
+    val map = mapOf<String, String>("Ы" to "И", "ы" to "и", "Я" to "А", "я" to "а", "Ю" to "У", "ю" to "у")
+    for (line in File(inputName).readLines())
+        if (line.length <= 1) outputStream.write("$line\n") else {
```
`else` с новой строки, `outputStream.write` взять в фигурные скобки

mglukhikh commented on src/lesson8/task1/Files.kt:
```diff
@@ -72,9 +72,21 @@ fun countSubstrings(inputName: String, substrings: List<String>): Map<String, In
+    for (line in File(inputName).readLines())
+        if (line.length <= 1) outputStream.write("$line\n") else {
+            outputStream.write(line[0].toString())
+            for (char in 1..line.length - 1)
```
Опять отсутствуют фигурные скобки

mglukhikh commented on src/lesson8/task1/Files.kt:
```diff
@@ -72,9 +72,21 @@ fun countSubstrings(inputName: String, substrings: List<String>): Map<String, In
+            for (char in 1..line.length - 1)
+                if ((line[char] in "ЫЯЮыяю") && (line[char - 1] in "ЖЧШЩжчшщ"))
+                    outputStream.write(map[line[char].toString()])
+                else outputStream.write(line[char].toString())
```
`outputStream.write` с новой строки -- как в основной ветке `if`

mglukhikh commented on src/lesson8/task1/Files.kt:
```diff
@@ -72,9 +72,21 @@ fun countSubstrings(inputName: String, substrings: List<String>): Map<String, In
+                if ((line[char] in "ЫЯЮыяю") && (line[char - 1] in "ЖЧШЩжчшщ"))
+                    outputStream.write(map[line[char].toString()])
+                else outputStream.write(line[char].toString())
+            outputStream.write("\n")
```
Здесь лучше использовать `newLine`, так как в разных ОС признаки перехода на новую строку разные

mglukhikh reviewed (commented):
Замечания, пожалуйста, поправьте.

mglukhikh commented:
Обнаружены точные копии решений других студентов. Решение о выставлении оценки "Отлично" откладывается до выяснения обстоятельств.

mglukhikh commented:
Этот PR я закрываю, исправления меня более не интересуют. Увидимся на экзамене.
